indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/PartitionLocation.java

```diff
@@ -25,36 +25,37 @@
 import org.joda.time.Interval;
 
 import java.net.URI;
+import java.util.Objects;
 
 /**
- * This class represents the intermediary data server where the partition of {@link #interval} and {@link #partitionId}
- * is stored.
+ * This class represents the intermediary data server where the partition of {@link #interval} and
+ * {@link #getPartitionId()} is stored.
  */
-public class PartitionLocation
+abstract class PartitionLocation<T>
 {
   private final String host;
   private final int port;
   private final boolean useHttps;
   private final String subTaskId;
   private final Interval interval;
-  private final int partitionId;
+  private final T secondaryPartition;
```

Comment:
What would be `secondaryPartition` for range partitioning?

Reply:
It uses the `ShardSpec`, so that it's generic enough for any kind of partitioning. Here's a preview of `GenericPartitionLocation`: https://github.com/ccaominh/incubator-druid/blob/backup-superbatch-range-partitioning/indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/GenericPartitionLocation.java

For has partitions, the existing implementation is more optimal with regards to having a smaller payload (since the partition dimensions do not need to be repeated for each partition).

Reply:
Hmm, does it need the entire shardSpec for range partitioning? I thought a triple of (startKey, endKey, partitionId) would be enough.

Reply:
The discussion of what exactly is used is probably best on the follow up PR that adds the implementation of range partitioning.

Reply:
Sounds right.